File: dotnet_backend/backend/Models/Batch.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace ComputerSeekho.Models;

/// <summary>
/// A cohort of one course.
/// </summary>
[Table("batches")]
public class Batch
{
    [Key]
    [Column("batch_id")]
    public int BatchId { get; set; }

    [Column("course_id")]
    public int CourseId { get; set; }

    /// <summary>Faculty leading the batch. NOT NULL in the schema.</summary>
    [Column("staff_id")]
    public int StaffId { get; set; }

    [Column("batch_name")]
    [Required, MaxLength(100)]
    public string BatchName { get; set; } = string.Empty;

    [Column("academic_year")]
    [MaxLength(20)]
    public string? AcademicYear { get; set; }

    [Column("start_date")]
    public DateOnly? StartDate { get; set; }

    [Column("end_date")]
    public DateOnly? EndDate { get; set; }

    [Column("presentation_date")]
    public DateTime? PresentationDate { get; set; }

    [Column("timing")]
    [MaxLength(100)]
    public string? Timing { get; set; }

    [Column("capacity")]
    public int Capacity { get; set; } = 20;

    /// <summary>
    /// A cached figure, recomputed from enrolments on every write and never
    /// accepted from a form. A counter that drifts once stays wrong.
    /// </summary>
    [Column("current_count")]
    public int CurrentCount { get; set; }

    [Column("status")]
    public BatchStatus Status { get; set; } = BatchStatus.Upcoming;

    [Column("is_active")]
    public bool IsActive { get; set; } = true;

    [ForeignKey(nameof(CourseId))]
    public Course? Course { get; set; }

    [ForeignKey(nameof(StaffId))]
    public Staff? Staff { get; set; }
}

File: backend/src/main/java/com/smvita/computerseekho/entity/Batch.java
package com.smvita.computerseekho.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDate;
import java.time.LocalDateTime;

@Entity
@Table(name = "batches")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class Batch {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "batch_id")
    private Integer batchId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "course_id", nullable = false)
    private Course course;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "staff_id", nullable = false)
    private Staff staff; // actual teaching faculty for this batch

    @Column(name = "batch_name", nullable = false, length = 100)
    private String batchName;

    @Column(name = "academic_year", length = 20)
    private String academicYear;

    @Column(name = "start_date")
    private LocalDate startDate;

    @Column(name = "end_date")
    private LocalDate endDate;

    @Column(name = "presentation_date")
    private LocalDateTime presentationDate;

    @Column(name = "timing", length = 100)
    private String timing;

    @Column(name = "capacity", nullable = false)
    private Integer capacity = 20;

    // System-calculated from enrollments — never hand-edited (see schema.sql comment).
    @Column(name = "current_count", nullable = false)
    private Integer currentCount = 0;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false,
            columnDefinition = "ENUM('Upcoming','Ongoing','Completed','Cancelled')")
    private Status status = Status.Upcoming;

    @Column(name = "is_active", nullable = false)
    private Boolean isActive = true;

    public enum Status { Upcoming, Ongoing, Completed, Cancelled }
}
